ports/stm32/mboot/main.c

```diff
@@ -731,9 +952,227 @@ int do_write(uint32_t addr, const uint8_t *src8, size_t len) {
+
+        } else if (addr == buff_next) {
+            // Next normal chunk into buffer
+            memcpy(&fw_enc_buff.data[addr-buff_addr-sizeof(chunk_header_v1_t)], src8, len);
+            fw_enc_buff_next = addr + len;
+
+        } else {
+            // New addr range, process existing buffer
+            // Copy new data into buffer afterwards
+            fw_enc_buff_addr = 0;
+        }
+
+        size_t fw_enc_buff_len = fw_enc_buff_next - buff_addr;
+
+        if (fw_enc_buff.header.header_vers != 1) {
+            dfu_state.status = DFU_STATUS_ERROR_FILE;
+            dfu_state.error = MBOOT_ERROR_STR_INVALID_SIG_IDX;
+            return -1;
+        }
+        if (fw_enc_buff.header.address != buff_addr) {
+            // Chunk address doesn't agree with dfu address, abort
+            dfu_state.status = DFU_STATUS_ERROR_ADDRESS;
+            dfu_state.error = MBOOT_ERROR_STR_INVALID_SIG_IDX;
+            return -1;
+        }
+
+        if (fw_enc_buff.header.length > fw_enc_buff_len) {
+            // Don't have entire chunk yet
+            return 0;
+        }
+
+        if (hydro_sign_verify(fw_enc_buff.header.signature, fw_enc_buff.data, fw_enc_buff.header.length, ENC_CONTEXT, *mboot_public_key) != 0) {
+            // Signature failed
+            if (fw_enc_buff_addr == 0) { 
+                // We should have a full chunk here, if it fails we abort entire process
+                dfu_state.status = DFU_STATUS_ERROR_VERIFY;
+                dfu_state.error = MBOOT_ERROR_STR_INVALID_SIG_IDX;
+                return -1;
+            } else {
+                return 0;
+            }
+        } else {
+            // Signature passes, we have valid message to flash
+            if (hydro_secretbox_decrypt(decrypted, fw_enc_buff.data, fw_enc_buff.header.length, 0, ENC_CONTEXT, *mboot_public_key) != 0) {
+                dfu_state.status = DFU_STATUS_ERROR_VERIFY;
+                dfu_state.error = MBOOT_ERROR_STR_INVALID_SIG_IDX;
+                return -1;
+            }
+
+            if (fw_enc_buff.header.format == GZIP) {
+                gzip = true;
+            }
+
+            // Use the decrypted message contents going formward
+            len = fw_enc_buff.header.length - hydro_secretbox_HEADERBYTES;
+            src8 = decrypted;
+            addr = buff_addr;
+        }
+
+        // We've just validated the previous buffer and a decrypted copy is about to be flashed
+        // Reset the buffer and copy this new chunk into it.
+        if (fw_enc_buff_addr == 0) {
+            fw_enc_buff_addr = addr;
+            memcpy(&fw_enc_buff.data[0], src8, len);
+            fw_enc_buff_next = addr + len;
+        }
+
+        // If we get here it means we've got a verified and decrypted chunk ready to write.
+        // Reset buffer marker for next chunk
+        fw_enc_buff_addr = 0;
+    }
+
+    uint8_t ret = 0;
+    if (gzip) {
+        TINF_DATA tinf = {0};
+        uint8_t buf[512*8];
```

Comment:
Must these numbers be fixed? What defines the size of the buffer? Best to describe the constants with variable names.